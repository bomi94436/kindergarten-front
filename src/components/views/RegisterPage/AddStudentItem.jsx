import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import IconButton from "@material-ui/core/IconButton";
import TextField from "@material-ui/core/TextField";
import "date-fns";
import DateFnsUtils from "@date-io/date-fns";
import {
  MuiPickersUtilsProvider,
  KeyboardDatePicker,
} from "@material-ui/pickers";
import { HiMinusCircle } from "react-icons/hi";
import { FaSearch } from "react-icons/fa";

import SearchKindergartenDialog from "./SearchKindergartenDialog";

const useStyles = makeStyles((theme) => ({
  row: {
    flexFlow: "row",
    justifyContent: "space-between",
    alignItems: "center",
  },
  input: {
    marginRight: "1rem",
  },
}));

const AddStudentItem = ({
  index,
  student,
  search,
  setRegister,
  setRegisterValid,
  setRegisterSearch,
  getRegisterSearch,
}) => {
  const classes = useStyles();

  const handleOpen = () => {
    setRegisterSearch({ name: "opened", value: true });
  };

  return (
    <>
      <Grid container className={classes.row}>
        {/* 이름 */}
        <Grid item xs={3}>
          <TextField
            variant="outlined"
            type="text"
            size="small"
            value={student.value.name || ""}
            onChange={(event) =>
              setRegister({
                name: "name",
                value: event.target.value,
                actor: "student",
                index: index,
              })
            }
            className={classes.input}
          />
        </Grid>

        {/* 생년월일 */}
        <Grid item xs={4} style={{ display: "flex" }}>
          <MuiPickersUtilsProvider utils={DateFnsUtils}>
            <KeyboardDatePicker
              format="yyyy/MM/dd"
              value={student.value.date}
              onChange={(date) =>
                setRegister({
                  name: "date",
                  value: date,
                  actor: "student",
                  index: index,
                })
              }
              className={classes.input}
            />
          </MuiPickersUtilsProvider>
        </Grid>

        {/* 소속 유치원 */}
        <Grid item xs={4} style={{ display: "flex" }}>
          <TextField
            variant="outlined"
            type="text"
            size="small"
            value={student.kindergarten_selected.name || ""}
            disabled
          />
          <IconButton
            type="button"
            variant="contained"
            color="primary"
            size="small"
            style={{ margin: "0.2rem", width: 33 }}
            onClick={handleOpen}
          >
            <FaSearch />
          </IconButton>
          <SearchKindergartenDialog
            search={search}
            setRegisterSearch={setRegisterSearch}
            getRegisterSearch={getRegisterSearch}
            actor="student"
            index={index}
          />
        </Grid>

        {/* 삭제 버튼 */}
        <Grid
          item
          xs={1}
          style={{
            display: "flex",
            justifyContent: "center",
            margin: "0.2rem",
          }}
        >
          <IconButton
            type="button"
            variant="contained"
            color="secondary"
            onClick={() =>
              setRegisterValid({
                actor: "user",
                act: "delete",
                index: index,
              })
            }
          >
            <HiMinusCircle />
          </IconButton>
        </Grid>
      </Grid>
    </>
  );
};

export default AddStudentItem;
