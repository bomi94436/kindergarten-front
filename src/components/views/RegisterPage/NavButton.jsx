import React from "react";
import { Link } from "react-router-dom";

const NavButton = () => {
  return (
    <div>
      <Link to="/register/parent">
        <button>학부모(일반회원)로 회원가입</button>
      </Link>
      <br />
      <Link to="/register/teacher">
        <button>선생님으로 회원가입</button>
      </Link>
      <br />
      <Link to="/register/director">
        <button>원장님으로 회원가입</button>
      </Link>
      <br />
      이미 계정이 있으신가요? <Link to="/login">로그인</Link>
    </div>
  );
};

export default NavButton;
