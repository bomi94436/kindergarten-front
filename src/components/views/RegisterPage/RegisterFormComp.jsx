import React from "react";
import styled from "styled-components";
import { makeStyles } from "@material-ui/core/styles";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import FormControl from "@material-ui/core/FormControl";
import InputLabel from "@material-ui/core/InputLabel";
import Select from "@material-ui/core/Select";
import MenuItem from "@material-ui/core/MenuItem";

import { inputIdByValid, inputPropsByValid } from "../../../utils/validation";

const useStyles = makeStyles((theme) => ({
  formControl: {
    margin: theme.spacing(1),
    minWidth: 120,
  },
}));

const InputCover = styled(TextField)`
  flex-grow: 1;
`;

export const FormInput = ({ element, setRegister, state }) => (
  <InputCover
    id={`${element.id} ${inputIdByValid(state.valid[element.id])}`}
    variant="outlined"
    error={inputPropsByValid(state.valid[element.id])}
    type={element.type}
    placeholder={element.placeholder}
    label={element.title}
    onChange={(event) =>
      setRegister({ name: element.id, value: event.target.value })
    }
    value={state.value[element.id]}
  />
);

export const FormButton = ({ element, getExistId, id, isUseridValid }) => (
  <Button
    id={element.id}
    type="button"
    variant="contained"
    color="primary"
    onClick={(event) => {
      if (!isUseridValid) {
        alert(
          "유효한 아이디를 입력해주세요.\n(영어, 숫자 조합하여 6자리 이상)"
        );
      } else {
        getExistId(id).then((res) => {
          alert(res.data);
        });
      }
    }}
  >
    {element.title}
  </Button>
);

export const FormDropdown = ({ element, setRegister, state }) => {
  const classes = useStyles();
  let value = state.value.lastEmail;
  return (
    <FormControl variant="outlined" className={classes.formControl}>
      <InputLabel id="dropdown">주소</InputLabel>
      <Select
        labelId="dropdown"
        id={element.id}
        onChange={(event) =>
          setRegister({
            name: element.id,
            value: event.target.value,
          })
        }
        value={value === "" ? "직접 입력" : value}
        label="주소"
      >
        {element.list.map((e, index) => (
          <MenuItem key={index} value={e}>
            {e}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};
