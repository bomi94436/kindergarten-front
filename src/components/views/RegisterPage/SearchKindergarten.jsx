import React from "react";

const SearchKindergarten = () => {
  return <div>SearchKindergarten</div>;
};

export default SearchKindergarten;
