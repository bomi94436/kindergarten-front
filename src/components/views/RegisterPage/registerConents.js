export const contents = {
  userid: {
    html: "input",
    id: "userid",
    title: "아이디",
    type: "text",
    placeholder: "숫자, 영어 조합하여 6자리 이상 입력",
  },
  checkDuplication: {
    html: "button",
    id: "checkDuplication",
    title: "중복 확인",
    type: "button",
  },
  password: {
    html: "input",
    id: "password",
    title: "비밀번호",
    type: "password",
    placeholder: "숫자, 영어 조합하여 8자리 이상 입력",
  },
  rePassword: {
    html: "input",
    id: "rePassword",
    title: "비밀번호 확인",
    type: "password",
    placeholder: "",
  },
  name: {
    html: "input",
    id: "name",
    title: "이름",
    type: "text",
    placeholder: "2글자 이상 입력",
  },
  phone: {
    html: "input",
    id: "phone",
    title: "휴대폰 번호",
    type: "number",
    placeholder: "숫자만 입력",
  },
  firstEmail: {
    html: "input",
    id: "firstEmail",
    title: "이메일",
    type: "text",
    placeholder: "",
  },
  lastEmail: {
    html: "input",
    id: "lastEmail",
    title: "",
    type: "text",
    placeholder: "",
  },
  emailList: {
    html: "dropdown",
    id: "emailList",
    list: ["직접 입력", "naver.com", "daum.net", "gmail.com"],
  },
};
