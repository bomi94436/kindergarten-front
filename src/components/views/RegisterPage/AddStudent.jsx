import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import Grid from "@material-ui/core/Grid";
import IconButton from "@material-ui/core/IconButton";

import { FaPlus } from "react-icons/fa";
import AddStudentItem from "../../../containers/RegisterPage/AddStudentItemContainer";

const useStyles = makeStyles((theme) => ({
  root: {
    flexFlow: "column",
    marginTop: "2rem",
    flexBasis: "100%",
  },
  row: {
    flexFlow: "row",
    justifyContent: "space-between",
    alignItems: "center",
  },
  label: {
    color: theme.palette.text.secondary,
    margin: "0.2rem",
  },
}));

const AddStudent = ({ students, setRegisterValid }) => {
  const classes = useStyles();

  return (
    <>
      <div className={classes.row} style={{ marginTop: "1rem" }}>
        <span style={{ margin: "1rem", fontSize: "1.2rem" }}>학생 추가</span>
        <IconButton
          type="button"
          variant="contained"
          color="primary"
          onClick={() => {
            setRegisterValid({
              actor: "user",
              act: "insert",
            });
          }}
        >
          <FaPlus />
        </IconButton>
      </div>

      <div className={classes.root}>
        <Grid container className={classes.row}>
          <Grid item xs={3}>
            <span className={classes.label}>이름</span>
          </Grid>
          <Grid item xs={4}>
            <span className={classes.label}>생년월일</span>
          </Grid>
          <Grid item xs={4}>
            <span className={classes.label}>소속 유치원</span>
          </Grid>
          <Grid item xs={1}>
            <span className={classes.label}>삭제</span>
          </Grid>
        </Grid>

        <p style={{ borderBottom: "1px solid gray", margin: "0.5rem" }}></p>

        {students.map((element, index) => {
          return <AddStudentItem key={index} index={index} student={element} />;
        })}
      </div>
    </>
  );
};

export default AddStudent;
