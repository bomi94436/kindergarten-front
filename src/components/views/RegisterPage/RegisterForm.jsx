import React, { useEffect, useState } from "react";
import SearchKindergarten from "./SearchKindergarten";
import { contents } from "./reigsterContents";

const RegisterForm = ({ history }) => {
  const [title, setTitle] = useState("");

  useEffect(() => {
    const path = history.location.pathname;
    if (path.indexOf("parent") !== -1) {
      setTitle("학부모(일반회원)");
    } else if (path.indexOf("teacher") !== -1) {
      setTitle("선생님");
    } else if (path.indexOf("director") !== -1) {
      setTitle("원장님");
    }
  }, []);

  const handleSubmit = (e) => {
    e.preventDefault();
  };

  return (
    <React.Fragment>
      <h1>
        {title}
        <br />
        회원가입
      </h1>
      <form onSubmit={(e) => handleSubmit(e)}>
        {contents.map((element) => {
          let content;
          let br = <br />;

          if (element.html === "input") {
            if (element.id === "email") {
              br = <span>@</span>;
            } else if (
              element.id === "userId" ||
              element.id === "emailDetail"
            ) {
              br = null;
            }

            content = (
              <React.Fragment key={element.id}>
                <label htmlFor={element.id}>{element.title}</label>
                <input
                  id={element.id}
                  type={element.type}
                  placeholder={element.placeholder}
                />
                {br}
              </React.Fragment>
            );
          } else if (element.html === "button") {
            content = (
              <React.Fragment key={element.id}>
                <button id={element.id}>{element.title}</button>
                {br}
              </React.Fragment>
            );
          } else if (element.html === "dropdown") {
            content = (
              <React.Fragment key={element.id}>
                <select id={element.id}>
                  {element.list.map((e, index) => (
                    <option key={index} value={e}>
                      {e}
                    </option>
                  ))}
                </select>
                {br}
              </React.Fragment>
            );
          }

          return content;
        })}

        {title === "선생님" || title === "원장님" ? (
          <SearchKindergarten />
        ) : null}

        <button type="submit">회원가입</button>
      </form>
    </React.Fragment>
  );
};

export default RegisterForm;
