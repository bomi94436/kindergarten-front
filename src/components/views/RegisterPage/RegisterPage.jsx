import React from "react";
import { Route, Switch } from "react-router-dom";
import NavButton from "./NavButton";
import RegisterForm from "./RegisterForm";

const RegisterPage = () => {
  return (
    <div>
      <Switch>
        <Route exact path="/register" component={NavButton} />
        <Route path="/register/parent" component={RegisterForm} />
        <Route path="/register/teacher" component={RegisterForm} />
        <Route path="/register/director" component={RegisterForm} />
      </Switch>
    </div>
  );
};

export default RegisterPage;
