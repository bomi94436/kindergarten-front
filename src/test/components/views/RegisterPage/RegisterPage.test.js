import React from "react";
import { Router } from "react-router-dom";
import { render, fireEvent } from "@testing-library/react";
import { createMemoryHistory } from "history";
import RegisterPage from "components/views/RegisterPage/RegisterPage";

describe("<RegisterPage />에서", () => {
  const history = createMemoryHistory();

  const setup = () => {
    history.push("/register");
    return render(
      <Router history={history}>
        <RegisterPage />
      </Router>
    );
  };

  it("잘 렌더링된다", () => {
    const component = setup();
  });

  it("스냅샷이 일치한다", () => {
    const component = setup();
    expect(component.container).toMatchSnapshot();
  });

  describe("회원가입 버튼 중에서, ", () => {
    it("학부모(일반회원)로 회원가입 버튼을 클릭했을 때, 잘 이동된다", () => {
      const component = setup();
      fireEvent.click(component.getByText("학부모(일반회원)로 회원가입"));
      expect(history.location.pathname).toBe("/register/parent");
    });

    it("선생님으로 회원가입 버튼을 클릭했을 때, 잘 이동된다", () => {
      const component = setup();
      fireEvent.click(component.getByText("선생님으로 회원가입"));
      expect(history.location.pathname).toBe("/register/teacher");
    });

    it("원장님으로 회원가입 버튼을 클릭했을 때, 잘 이동된다", () => {
      const component = setup();
      fireEvent.click(component.getByText("원장님으로 회원가입"));
      expect(history.location.pathname).toBe("/register/director");
    });
  });
});
