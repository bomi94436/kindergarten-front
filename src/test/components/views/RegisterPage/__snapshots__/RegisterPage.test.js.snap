// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterPage />에서 스냅샷이 일치한다 1`] = `
<div>
  <div>
    <div>
      <a
        href="/register/parent"
      >
        <button>
          학부모(일반회원)로 회원가입
        </button>
      </a>
      <br />
      <a
        href="/register/teacher"
      >
        <button>
          선생님으로 회원가입
        </button>
      </a>
      <br />
      <a
        href="/register/director"
      >
        <button>
          원장님으로 회원가입
        </button>
      </a>
      <br />
      이미 계정이 있으신가요? 
      <a
        href="/login"
      >
        로그인
      </a>
    </div>
  </div>
</div>
`;
